src/pywrangler/sync.py

```diff
@@ -0,0 +1,267 @@
+PYPROJECT_TOML_PATH = PROJECT_ROOT / "pyproject.toml"
+GENERATED_REQUIREMENTS_PATH = PYODIDE_VENV_PATH / "temp-requirements.txt"
+VENV_REQUIREMENTS_PATH = VENV_WORKERS_PATH / "temp-venv-requirements.txt"
+
+
+def check_pyproject_toml():
+    if not PYPROJECT_TOML_PATH.is_file():
+        logger.error(f"{PYPROJECT_TOML_PATH} not found.")
+        raise click.exceptions.Exit(code=1)
+
+
+def create_workers_venv():
+    """
+    Creates a virtual environment at `VENV_WORKERS_PATH` if it doesn't exist.
+    """
+    if not VENV_WORKERS_PATH.is_dir():  # Check if it's a directory
+        logger.debug(f"Creating virtual environment at {VENV_WORKERS_PATH}...")
+        run_command(["uv", "venv", str(VENV_WORKERS_PATH), "--python", "python3.12"])
+    else:
+        logger.debug(f"Virtual environment at {VENV_WORKERS_PATH} already exists.")
+
+
+def _get_pyodide_cli_path():
+    venv_bin_path = VENV_WORKERS_PATH / ("Scripts" if os.name == "nt" else "bin")
+    pyodide_cli_path = venv_bin_path / ("pyodide.exe" if os.name == "nt" else "pyodide")
+    return pyodide_cli_path
+
+
+def install_pyodide_build():
+    pyodide_cli_path = _get_pyodide_cli_path()
+
+    if not pyodide_cli_path.is_file():
+        logger.debug(
+            f"Installing pyodide-build in {VENV_WORKERS_PATH} using 'uv pip install'..."
+        )
+        venv_bin_path = pyodide_cli_path.parent
+
+        # Ensure the python executable path is correct for the venv
+        venv_python_executable = venv_bin_path / "python"
+        if not venv_python_executable.is_file():
+            logger.error(f"Python executable not found at {venv_python_executable}")
+            raise click.exceptions.Exit(code=1)
+
+        run_command(["uv", "pip", "install", "-p", str(venv_python_executable), "pip"])
+
+        run_command(
+            ["uv", "pip", "install", "-p", str(venv_python_executable), "pyodide-build"]
+        )
+    else:
+        logger.debug(
+            f"pyodide-build CLI already found at {pyodide_cli_path} (skipping install.)"
+        )
+
+
+def create_pyodide_venv():
+    pyodide_cli_path = _get_pyodide_cli_path()
+    if PYODIDE_VENV_PATH.is_dir():
+        logger.debug(
+            f"Pyodide virtual environment at {PYODIDE_VENV_PATH} already exists."
+        )
+        return
+
+    logger.debug(f"Creating Pyodide virtual environment at {PYODIDE_VENV_PATH}...")
+    PYODIDE_VENV_PATH.parent.mkdir(parents=True, exist_ok=True)
+    run_command([str(pyodide_cli_path), "venv", str(PYODIDE_VENV_PATH)])
+
+
+def generate_requirements() -> bool:
+    logger.debug(
+        f"Reading dependencies from {PYPROJECT_TOML_PATH} and generating {GENERATED_REQUIREMENTS_PATH}..."
+    )
+    try:
+        with open(PYPROJECT_TOML_PATH, "rb") as f:  # Use binary mode for tomllib
+            pyproject_data = tomllib.load(f)
+
+        # Extract dependencies from [project.dependencies]
+        dependencies = pyproject_data.get("project", {}).get("dependencies", [])
```

Comment:
How about just returning the dependencies here? And then we can put `@functools.cache` around this? No need to create a file.